### PR TITLE
Using a Form Object that will persist for the Creator Settings Form

### PR DESCRIPTION
**Pull request by Ridhwana**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description

Demo video can be found [here](https://www.loom.com/share/1e9100f4dd714c3892f00aabac8dffc9).

The problem that we had on this page and on many other pages in the Admin, is that once a user submits their data in a form, and if it errors out, it shows a flash of an error and then a blank form, forcing users to start over. 

In this scenario, on the Creator Settings Page, we can actually save the transactions that have passed as they are not dependant on each other as we previously thought. Hence, when we submit a form, and something fails, we want to show an error message stating what has failed and also display the information that the user has filled in that has already been saved to the database.

In addition, we have set default values on `Settings` class that we should show as defaults on the form. 

After chatting to Michale Kohl, we we decided to use Form Objects for this -  it cleans up a lot of aspects moves conditional callbacks and validations out of models so we don’t have to worry about accidentally triggering them from unrelated actions. In addition, it provides with a good way to update the multiple Settings at the same time.

You can read more about Form Object [here](https://nimblehq.co/blog/lets-play-design-patterns-form-objects)

## Related Tickets & Documents
Closes https://github.com/forem/forem/issues/15651

## QA Instructions, Screenshots, Recordings

For this PR its best to drop your db, migrate and enable your feature flag to test as a new user with the default Settings Applied.

```
rails db:drop RAILS_ENV=development
rails db:create RAILS_ENV=development
rails db:migrate RAILS_ENV=development

FeatureFlag.enable(:creator_onboarding)
```

Now, run your server and navigate to http://localhost:3000/ and complete the "Create Account" form. After completing the "Create Account" form, you should be brought to the new Creator Setup page, ${baseUrl}/admin/creator_settings/new?referrer=${baseUrl}.

You can play around with the form and make sure that the ffg values are saved as per you update. FIll out your values on the form and then chcek them in the rails console: 

```
Settings::Community.community_name
Settings::UserExperience.primary_brand_color_hex
Settings::UserExperience.public
Settings::Authentication.invite_only_mode
Settings::General.original_logo
Settings::General.resized_logo
```

To mimic an error, and then test the other values get saves besides the logo you can save an image thats dimensions are greater than 4096px x 4096px as per the demo. 
 

### UI accessibility concerns?

None, there were no frontend changes.

## Added/updated tests?

- [x] Yes
- [ ] No, and this is why: _please replace this line with details on why tests
      have not been included_
- [ ] I need help with writing tests

## [Forem core team only] How will this change be communicated?

_Will this PR introduce a change that impacts Forem members or creators, the
development process, or any of our internal teams? If so, please note how you
will share this change with the people who need to know about it._

- [ ] I've updated the [Developer Docs](https://developers.forem.com) or
      [Storybook](https://storybook.forem.com/) (for Crayons components)
- [ ] This PR changes the Forem platform and our documentation needs to be
      updated. I have filled out the
      [Changes Requested](https://github.com/forem/admin-docs/issues/new?assignees=&labels=&template=changes_requested.md)
      issue template so Community Success can help update the Admin Docs
      appropriately.
- [ ] I've updated the README or added inline documentation
- [ ] I've added an entry to
      [`CHANGELOG.md`](https://github.com/forem/forem/tree/main/CHANGELOG.md)
- [ ] I will share this change in a [Changelog](https://forem.dev/t/changelog)
      or in a [forem.dev](http://forem.dev) post
- [ ] I will share this change internally with the appropriate teams
- [ ] I'm not sure how best to communicate this change and need help
- [x] This change does not need to be communicated now and will be communicated when the feature is released.

## [optional] What gif best describes this PR or how it makes you feel?

alt_text

### CODE REVIEW

File: app/forms/creator_settings_form.rb [hunk 1]
```diff
@@ -13,16 +13,14 @@ class CreatorSettingsForm
...
-    @primary_brand_color_hex = primary_brand_color_hex
-    @invite_only_mode = invite_only_mode
-    @public = public
-    @logo = logo
-    @checked_code_of_conduct = checked_code_of_conduct
-    @checked_terms_and_conditions = checked_terms_and_conditions
+  def initialize(attributes = {})
+    super
+
+    @primary_brand_color_hex = primary_brand_color_hex || ::Settings::UserExperience.primary_brand_color_hex
```
Comment (citizen428) [inline app/forms/creator_settings_form.rb:19]: If this is setting the attributes defined above via `attribute`, maybe use the following form instead:

```ruby
self.primary_brand_color_hex = ...
```

File: app/forms/creator_settings_form.rb [hunk 2]
```diff
@@ -13,16 +13,14 @@ class CreatorSettingsForm
...
+    @invite_only_mode = invite_only_mode || 1
+    @public = public || 1
```
Comment (citizen428) [inline app/forms/creator_settings_form.rb:21]: The `public` here clearly confuses the syntax highlighter but it should work ok as Ruby allows for overriding these names (though it may still be confusing for the reader).

File: spec/forms/creator_settings_form_spec.rb [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+        checked_code_of_conduct: true,
+        checked_terms_and_conditions: true,
+        community_name: "Climbing Life",
+        invite_only_mode: false,
+        public: false,
+        primary_brand_color_hex: "#a81adb",
+      )
+
+      expect(creator_settings_form.valid?).to be(true)
+      expect(Settings::Community.community_name).to eq("Climbing Life")
```
Comment (Ridhwana) [inline spec/forms/creator_settings_form_spec.rb:47]: @citizen428 I was wondering if you had some insights on this test that I'm struggling with please. This test block (with the Settings updates) are failing.

The code in the console is working as expected i.e. When we have a valid form object and we call `.save` on it, it updates all the Settings values. I see the expected results in my rails console if I test as follows:

```
c = CreatorSettingsForm.new(
        checked_code_of_conduct: true,
        checked_terms_and_conditions: true,
        community_name: "Climbing Life",
        invite_only_mode: false,
        public: false,
        primary_brand_color_hex: "#a81adb",
      )

c.valid?
# returns true

c.save
```

Now if I check the Settings values:
```
Settings::UserExperience.primary_brand_color_hex
# returns "a81adb"

Settings::UserExperience.public
# returns true
```

However, when I run this spec, the Settings values do not update. I know that on a model we'd usually do a `reload`, do you know how we'd check for the updated values in this case?

File: app/controllers/admin/creator_settings_controller.rb [hunk 4]
```diff
@@ -4,34 +4,33 @@ class CreatorSettingsController < Admin::ApplicationController
                         invite_only_mode logo primary_brand_color_hex public].freeze
 
     def new
+      @creator_settings_form = CreatorSettingsForm.new(
```
Comment (Ridhwana) [inline app/controllers/admin/creator_settings_controller.rb:7]: We send these parameters through as initial values because when we have an error, then we want the values that were set in the form to show up. We also want to pull through the default values for Settings.

File: app/controllers/admin/creator_settings_controller.rb [hunk 5]
```diff
@@ -4,34 +4,33 @@ class CreatorSettingsController < Admin::ApplicationController
...
+        primary_brand_color_hex: ::Settings::UserExperience.primary_brand_color_hex,
+      )
       @max_file_size = LogoUploader::MAX_FILE_SIZE
       @logo_allowed_types = (LogoUploader::CONTENT_TYPE_ALLOWLIST +
         LogoUploader::EXTENSION_ALLOWLIST.map { |extension| ".#{extension}" }).join(",")
     end
 
     def create
       extra_authorization
-      ActiveRecord::Base.transaction do
```
Comment (Ridhwana) [inline app/controllers/admin/creator_settings_controller.rb:14]: You'll notice that we completely remove `ActiveRecord::Base.transaction do`. I chatted to @citizen428 about this and we concluded that each of those Settings are not directly dependant on each other, so it may be okay to save some and to then allow them to update others afterwards if it fails.

File: app/controllers/admin/creator_settings_controller.rb [hunk 6]
```diff
@@ -4,34 +4,33 @@ class CreatorSettingsController < Admin::ApplicationController
...
-        ::Settings::UserExperience.primary_brand_color_hex = settings_params[:primary_brand_color_hex]
-        ::Settings::Authentication.invite_only_mode = settings_params[:invite_only]
-        ::Settings::UserExperience.public = settings_params[:public]
 
-        if settings_params[:logo]
-          logo_uploader = upload_logo(settings_params[:logo])
-          ::Settings::General.original_logo = logo_uploader.url
-          ::Settings::General.resized_logo = logo_uploader.resized_logo.url
-        end
+      @creator_settings_form = CreatorSettingsForm.new(settings_params)
```
Comment (Ridhwana) [inline app/controllers/admin/creator_settings_controller.rb:21]: Instead of handling this here, we call a `CreatorSettingForm` that moves conditional callbacks and validations into a plain old ruby object. This Ruby object handles the updates and validations.

File: app/forms/creator_settings_form.rb [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+        ::Settings::UserExperience.public = public
+
+        if logo
+          logo_uploader = upload_logo(logo)
+          ::Settings::General.original_logo = logo_uploader.url
+          ::Settings::General.resized_logo = logo_uploader.resized_logo.url
+        end
+        @success = true
+      rescue StandardError => e
+        errors.add(:base, e.message)
```
Comment (Ridhwana) [inline app/forms/creator_settings_form.rb:50]: Here we capture any other errors related to the code execution (and not only validation).

File: app/models/settings/community.rb [hunk 8]
```diff
@@ -10,7 +10,7 @@ class Community < Base
     setting(
       :community_name,
       type: :string,
-      default: ApplicationConfig["COMMUNITY_NAME"] || "New Forem",
+      default: ApplicationConfig["COMMUNITY_NAME"],
```
Comment (Ridhwana) [inline app/models/settings/community.rb:13]: We remove "New Forem", because when a creator signs up and then goes through their onboarding, we want to show them a blank field or what was set by their `ApplicationConfig["COMMUNITY_NAME"]`.

"New Forem" no longer makes sense with this onboarding step at the start.

File: app/models/settings/user_experience.rb [hunk 9]
```diff
@@ -21,7 +21,7 @@ class UserExperience < Base
     }
     # a non-public forem will redirect all unauthenticated pages to the registration page.
     # a public forem could have more fine-grained authentication (listings ar private etc.) in future
-    setting :public, type: :boolean, default: 0
+    setting :public, type: :boolean, default: 1
```
Comment (Ridhwana) [inline app/models/settings/user_experience.rb:24]: By default we want the form to be public. Refer to [slack conversation](https://forem-team.slack.com/archives/C01S4QZUDNC/p1638896683273800).

File: spec/forms/creator_settings_form_spec.rb (hunk 3 above)
Comment (citizen428) [reply spec/forms/creator_settings_form_spec.rb:46]: @Ridhwana Reading this it seems like you're not actually saving the form in the spec. So I added an explicit call to `creator_settings_form.save` after the `expect(creator_settings_form.valid?).to be(true)` assertion and that did fix all but one spec. Not sure why `public` is not updating yet.

File: app/forms/creator_settings_form.rb [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+  validates :checked_terms_and_conditions, inclusion: { in: [true, false] }
+  validates :invite_only_mode, inclusion: { in: [true, false] }
+  validates :public, inclusion: { in: [true, false] }
+
+  attr_accessor :success
+
+  def initialize(attributes = {})
+    super
+
+    self.checked_code_of_conduct = checked_code_of_conduct
```
Comment (citizen428) [inline app/forms/creator_settings_form.rb:26]: This confused me a bit. When you put a `binding.pry` after `super`, you'll see that all attributes on the left-hand side already have the value that's on the right hand side because you're using the getter to update the setter, which means you overwrite the value with itself. 

Completely removing the `initialize` method seems to work fine (the initializer spec is still green) because we end up using the one from `ActiveModel::Attributes`:

```ruby
CreatorSettingsForm.new.method(:initialize).owner
#=> ActiveModel::Attributes
```

File: app/models/settings/user_experience.rb (hunk 9 above)
Comment (citizen428) [reply app/models/settings/user_experience.rb:24]: We should probably change this to `true` instead of `1`, usually we don't use numeric values for booleans in settings.

**Review by citizen428 — APPROVED** [review]

**Comment by citizen428** [comment]
Sorry, I had accidentally approved instead of just commenting.

File: spec/forms/creator_settings_form_spec.rb (hunk 3 above)
Comment (citizen428) [reply spec/forms/creator_settings_form_spec.rb:46]: BTW, regarding the problem with `true`, that also shows in your console example above:

```ruby
...
public: false,

Settings::UserExperience.public
# returns true
```

Comment (citizen428) [reply spec/forms/creator_settings_form_spec.rb:51]: Found the culprit! In `./spec/rails_helper.rb` we have the following:

```ruby
    allow(Settings::UserExperience).to receive(:public).and_return(true)
```

This is a leftover from the profile migration because otherwise we'd have needed to change a lot of specs. 

I took the liberty of pushing the fix directly to this branch, hope that's ok.

File: app/models/settings/user_experience.rb [hunk 11]
```diff
@@ -21,7 +21,7 @@ class UserExperience < Base
     }
     # a non-public forem will redirect all unauthenticated pages to the registration page.
     # a public forem could have more fine-grained authentication (listings ar private etc.) in future
-    setting :public, type: :boolean, default: 0
+    setting :public, type: :boolean, default: true
```
Comment (Ridhwana) [inline app/models/settings/user_experience.rb:24]: Since the type here is boolean it makes set the value to a boolean too. In addition,  we've decided to set public to `true` based on [this](https://forem-team.slack.com/archives/C01S4QZUDNC/p1638896683273800) conversation

File: app/views/admin/creator_settings/new.html.erb [hunk 12]
```diff
@@ -7,37 +7,23 @@
...
-          <div class="crayons-card__body">
-            <ul>
-              <% resource.errors.full_messages.each do |message| %>
-                <li><%= message %></li>
-              <% end %>
-            </ul>
-          </div>
-        </div>
-      <% end %>
+    <%= form_with url: admin_creator_settings_path, model: @creator_settings_form, local: true, method: "post", multipart: true, class: "relative z-elevate p-4", "data-action": "submit->creator-settings#formValidations" do |f| %>
```
Comment (Ridhwana) [inline app/views/admin/creator_settings/new.html.erb:17]: We user the `creator_settings_form` model to render the form. In addition, now when there is an error, the values that were saved already pull through in the form.

**Comment by nickytonline** [comment]
@Ridhwana, I ran a quick test by uploading a non-image type, and the page shows the error in the flash message with the form data populated except for the checkboxes.

image

I'm going to dig a bit in the code to see what's up.

File: app/controllers/admin/creator_settings_controller.rb (hunk 6 above)
Comment (nickytonline) [reply app/controllers/admin/creator_settings_controller.rb:21]: @Ridhwana, I ran a quick test by uploading a non-image type, and the page shows the error in the flash message with the form data populated except for the checkboxes.

image

This is an example of the Ruby object being returned (I converted it to JSON)

```
{"attributes":{"checked_code_of_conduct":true,"checked_terms_and_conditions":true,"community_name":"ro","invite_only_mode":false,"logo":{"tempfile":"#\u003cFile:0x00007fe1a64a6cc8\u003e","original_filename":"divvy-invoice-2021.pdf","content_type":"application/pdf","headers":"Content-Disposition: form-data; name=\"creator_settings_form[logo]\"; filename=\"divvy-invoice-2021.pdf\"\r\nContent-Type: application/pdf\r\n"},"primary_brand_color_hex":"#3b49df","public":true}}
```

I see the checkbox values in there, but still digging to see why the checkbox values aren't being updated.

Comment (Ridhwana) [reply app/controllers/admin/creator_settings_controller.rb:21]: Ah @nickytonline I don't persist the terms and conditions checkboxes on the user object until all the fields pass. That made the most sense to me and I've seen it done this way before, but if that doesn't make sense to others, we can change it.

Comment (nickytonline) [reply app/controllers/admin/creator_settings_controller.rb:21]: I thought it was a bug, but I honestly don't know if that's a pattern for accepting a ToS and CoC. I trust your judgement on this one.

Comment (Ridhwana) [reply app/controllers/admin/creator_settings_controller.rb:21]: I just posted in our channel to get product weigh in :) Thanks for bringing it up!

File: spec/requests/admin/creator_settings_spec.rb [hunk 13]
```diff
@@ -55,7 +59,8 @@
         expect(creator.saw_onboarding).to eq(true)
         expect(creator.checked_code_of_conduct).to eq(true)
         expect(creator.checked_terms_and_conditions).to eq(true)
-        expect(response).to have_http_status(:ok)
+        # we redirect immadiately on success
+        expect(response).to have_http_status(:found)
```
Comment (maestromac) [inline spec/requests/admin/creator_settings_spec.rb:63]: You can have a test case that also explains itself.
```suggestion
expect(response).to redirect_to(:root).and have_http_status(:found)
```

**Comment by maestromac** [comment]
I'm trying to run this locally but when I head to http://localhost:3000/admin/creator_settings/new all I see is a blank page. Is there something I missed?

**Comment by Ridhwana** [comment]
> I'm trying to run this locally but when I head to http://localhost:3000/admin/creator_settings/new all I see is a blank page. Is there something I missed?

@maestromac I'm not sure why that would happen. I've not experienced this before. Can you tell me what steps you took before getting to the page? Also did you enable the creator_onboarding feature flag?

**Comment by maestromac** [comment]
@Ridhwana drop create migrate, enable feature flag in console.

**Comment by Ridhwana** [comment]
> @Ridhwana drop create migrate, enable feature flag in console.

Hmm that's peculiar. 
@nickytonline when you tested, did you follow the QA Steps and could you access the page?

**Review by maestromac — APPROVED** [review]
Tested locally. LGTM! pending conflict files

File: app/controllers/admin/creator_settings_controller.rb (hunk 6 above)
Comment (Ridhwana) [reply app/controllers/admin/creator_settings_controller.rb:21]: @nickytonline  I've gone ahead and persisted the toc and coc as per [the product discussion](https://forem-team.slack.com/archives/C01S4QZUDNC/p1639077971299600). Thanks once again for bringing this up!

screencapture-localhost-3000-admin-creator-settings-new-2021-12-10-11_37_22

File: app/assets/javascripts/initializers/initializePodcastPlayback.js [hunk 14]
```diff
@@ -475,7 +475,9 @@ function initializePodcastPlayback() {
 
   function handlePodcastMessages(event) {
     const message = JSON.parse(event.detail);
-    if (message.namespace !== 'podcast') { return }
```
Comment (Ridhwana) [inline app/assets/javascripts/initializers/initializePodcastPlayback.js:478]: unrelated eslint changes form master

File: app/javascript/onboarding/components/Navigation.jsx [hunk 15]
```diff
@@ -52,15 +52,8 @@ export class Navigation extends Component {
...
-    const {
-      next,
-      prev,
-      hideNext,
-      hidePrev,
-      disabled,
-      canSkip,
-      className,
-    } = this.props;
+    const { next, prev, hideNext, hidePrev, disabled, canSkip, className } =
```
Comment (Ridhwana) [inline app/javascript/onboarding/components/Navigation.jsx:55]: unrelated eslint changes when I pulled in master

**Review by juliannatetreault — APPROVED** [review]
Works like a charm! ✨
